Add per key actor epochs

russelldb opened the pull request:
See riak_kv#679 and the associated platform_task RFC and summary.

Requires https://github.com/basho/riak_core/pull/646

This PR addresses the "doomstone", backup-restore, and some byzantine flavours of the kv679 bug.

The RFC explains the mechanism in detail but briefly:
1. add a persisted to disk vnode counter (persisted with leases, aysnc)
2. when ever a key is written to for the "first time" by a vnode, create an epoch actor for the key by concatenating vnodeid+counter (and increment the counter)

This ensures that a first time write for a key gets a new actor, this is the epoch for the key. It means we don't mix up deleted+re-created keys {a,1} event with the original {a, 1} event for some key, by ensuring an actor per epoch, without causing a keyspace wide actor explosion.

I hope that is enough details for the review to start. I'll start work on EQC for the status_mgr and the new vnode functions in the mean time.

russelldb committed:
WIP move vnode persistent state into new module/process

russelldb committed:
WIP kv679 fixes. Adds a per vnode monotonic counter

use the counter to create a per epoch key when local not found.

russelldb committed:
WIP: don't allow the counter to get above 32bit int size

When it hits that limit create a new vnodeid.

russelldb committed:
WIP fix xref errors

russelldb committed:
WIP fix dialyzer

russelldb committed:
WIP really fix dialyzer

russelldb committed:
WIP only increment vnode counter on new per key epoch

Start a new epoch if the local object has a lower epoch for the
vnode id than the incoming object (no epoch is lower, no epich is zero)

russelldb committed:
WIP fix new dialyzer errors

Bad match for re-refactord put_merge return.
Correct return type for highest_actor

russelldb committed:
WIP update the vnode status command as per fix on 2.0 branch

russelldb committed:
Fix failing stats test by using correct vnode status format

russelldb committed:
Don't force a new epoch on old keys

use the bare vnodeid as long as possible, only start a new epoch
when needed. In a way the bare vnodeid is just epoch zero, so use it.

Also, consider the incoming counter when deciding about new epochs.
An incoming clock with the same actor+epoch but greater counter is
a hint that a byzantine failure occurred and a new epoch is needed

russelldb committed:
Fix some doco/spec errors, fix an error getting the timeout for leases

russelldb committed:
Shhhh, loud test.

russelldb committed:
Add some docs and specs and fix the transposed args that dialyzer found

NOTE: make dialyzer exits with an error, but no information. Could
do with some help on that.

russelldb committed:
fix filename spec typo

russelldb committed:
Add a `status` command for testing purposes

russelldb committed:
WIP for help on test

seancribbs commented on src/riak_kv_vnode.erl:
```diff
@@ -2286,102 +2297,202 @@ new_md_cache(VId) ->
+    State;
+maybe_lease_counter(State) ->
+    #state{status_mgr_pid=MgrPid, counter=CS=#counter_state{cnt=Cnt, lease=Lease, lease_size=LeaseSize}} = State,
+    %% @TODO configurable??
```
Yes, this should probably be configurable/tunable.

russelldb replied:
Yeah. It is hard to decide if it should be. Trying to figure out a scenario where you'd want to change this. And adding it adds more chance to shoot yourself in the foot. And more complex validation of the config item. And can it change in the vnode's lifetime? Do we need to add it to state, or check the env var every iteration?

I'm not against making it configurable, I'd just like a better understanding of the benefit vs. risk/complexity.

seancribbs commented on src/riak_kv_vnode.erl:
```diff
@@ -2286,102 +2297,202 @@ new_md_cache(VId) ->
+    #state{idx=Index, status_mgr_pid=Pid, counter=CounterState} = State,
+    #counter_state{lease_size=LeaseSize} = CounterState,
+    Start = os:timestamp(),
+    receive
```
Could this whole blocking paradigm be replaced with `gen_server:call` with a timeout? Also isn't it possible to enter this receive without having sent a lease request yet? Where is the original cast?

russelldb replied:
re gen_server call, no, I think that breaks the encapsulation and serialization of file access, and lays us open to races.
re entering receive without requesting a lease: how? The state for counter must be "leasing=true"

seancribbs commented on src/riak_kv_vnode.erl:
```diff
@@ -2286,102 +2297,202 @@ new_md_cache(VId) ->
+    #state{status_mgr_pid=MgrPid, counter=CS=#counter_state{cnt=Cnt, lease=Lease, lease_size=LeaseSize}} = State,
+    %% @TODO configurable??
+    %% has more than 80% of the lease been used?
+    CS2 = case (Lease - Cnt) =< 0.2 * LeaseSize of
```
All of the expressions in this are guard expressions, an `if` would be clearer in this case.

russelldb replied:
If you say so.

seancribbs commented on src/riak_kv_vnode.erl:
```diff
@@ -1484,11 +1553,19 @@ select_newest_content(Mult) ->
 
 %% @private
 put_merge(false, true, _CurObj, UpdObj, _VId, _StartTime) -> % coord=false, LWW=true
+    %% @TODO Do we need to mark the clock dirty here? I think so
```
Why do you need to mark the clock dirty here? LWW should ignore the clock.

russelldb replied:
But we allow people to change from LWW->OTHER, so a detected dirty clock should still be marked for future reads.

seancribbs commented on src/riak_kv_vnode.erl:
```diff
@@ -1368,6 +1435,8 @@ prepare_new_put(true, RObj, VId, StartTime, CRDTOp) ->
     %% Make a new crdt, stuff it in the riak_object
     do_crdt_update(VClockUp, VId, CRDTOp);
 prepare_new_put(false, RObj, _VId, _StartTime, _CounterOp) ->
+    %% @TODO Not coordindating, not found local, is there an entry for
```
Is the goal of marking the entry/clock dirty to ensure that it gets a new epoch next time this actor coordinates? If so, then :+1: to that. That seems like an early indicator that data was lost.

russelldb replied:
But this is future work, of course, as it requires further riak_object metadata.

seancribbs commented on src/riak_kv_vnode.erl:
```diff
@@ -1279,12 +1346,12 @@ prepare_put(#state{vnodeid=VId,
+                %% incoming object. If the incoming object descends
+                %% the cache (i.e. has seen all its events) no need to
+                %% do a local get and merge, just overwrite.
+                not riak_object:vclock_descends(RObj, Clock)
```
I love this clarification.

seancribbs commented on src/riak_kv_vnode.erl:
```diff
@@ -1174,18 +1239,18 @@ raw_put({Idx, Node}, Key, Obj) ->
-        false ->
-            PruneTime = StartTime
-    end,
+    BProps =  case proplists:get_value(bucket_props, Options) of
```
Thank you for removing the case-export warnings.

seancribbs commented on src/riak_kv_vnode.erl:
```diff
@@ -1134,8 +1176,31 @@ handle_info({final_delete, BKey, RObjHash}, State = #state{mod=Mod, modstate=Mod
+    #counter_state{lease_size=LeaseSize, leasing=Leasing} = CntrState,
+    {ok, NewPid} = riak_kv_vnode_status_mgr:start_link(self(), Index),
+
+    case Leasing of
```
This seems like a clear spot for `if` instead of `case`.

russelldb replied:
Personally _HATE_ erlang `if` especially in this case, as it mentally comes out as

   if true -> thing;
     true -> other_thing
    end

cmeiklejohn replied:
I'm with @russelldb.

seancribbs commented on src/riak_kv_vnode.erl:
```diff
@@ -1134,8 +1176,31 @@ handle_info({final_delete, BKey, RObjHash}, State = #state{mod=Mod, modstate=Mod
+handle_info({counter_lease, {FromPid, VnodeId, NewLease}}, State=#state{status_mgr_pid=FromPid}) ->
+    #state{counter=CounterState} = State,
+    #counter_state{lease=OldLease} = CounterState,
+    %% @TODO (rdb) Paranoia? If the NewLease < OldLease something is very broken,
```
We should discuss the implications of the NewLease being smaller than the OldLease. Seems like reordered messages could cause it (but are nigh-impossible between two local processes), but could dropped messages? Perhaps this needs a PULSE model.

russelldb replied:
I agree, let's discuss.

Does erlang  drop messages between local processes?

russelldb replied:
Actually this is very broken. The old lease can legitimately be smaller than the old lease. If we've gone over the `?MAX_INT` boundary and got a new vnodeid, then the new lease is expected to be smaller, and using `max` is wrong and broken, we'll be handing out counters that are not on disk.

seancribbs commented on src/riak_kv_vnode.erl:
```diff
@@ -2286,102 +2297,202 @@ new_md_cache(VId) ->
+%% and max errors, use the defaults, and log the insanity. Note this
+%% expects the macro configured defaults to be positive integers!
+-spec non_neg_env(atom(), atom(), pos_integer()) -> pos_integer().
+non_neg_env(App, EnvVar, Default) when is_integer(Default),
```
For the future, this seems like something that could be in `app_helper` in core.

russelldb replied:
Noted, happy to move

seancribbs commented on src/riak_kv_vnode.erl:
```diff
@@ -2286,102 +2297,202 @@ new_md_cache(VId) ->
+
+%% @private generate a new epoch ID for a key
+-spec key_epoch_actor(vnodeid(), pos_integer()) -> binary().
+key_epoch_actor(ActorBin, Cntr) ->
```
It seems you only call this from the function above. Do we need a separate abstraction for the construction of this ID? If we do, shouldn't we have a deconstructor too (probably used by `highest_actor()`)?

russelldb replied:
Hmmm. Ok.

russelldb replied:
So, we'd be tying ourselves to a set length for the `ActorBin` part, no? And we'd need versioning in case that changed in future. Maybe `highest_actor()` should use the length of `VnodeId` as the base to match from? Or can we trust the `VnodeId` is 8 bytes for ever and add the deconstructor as you suggest?

russelldb replied:
I changed this so that we don't know, nor need to know, that `vnodeid` is currently 8 bytes. This makes changing the `vnodeid` scheme down the road simpler, and keeps the matching backwards/forwards compatible.

seancribbs commented on src/riak_kv_vnode.erl:
```diff
@@ -2286,102 +2297,202 @@ new_md_cache(VId) ->
+     Counter :: non_neg_integer()}.
+highest_actor(ActorBase, Obj) ->
+    Actors = riak_object:all_actors(Obj),
+    {Actor, Epoch} = highest_actor(ActorBase, Actors, undefined, 0),
```
Like I mentioned on old commits, this feels like a fold since you aren't bailing early from the processing.  How about this refactor:

``` erlang
{Actor, Epoch} = lists:foldl(find_highest_actor(ActorBase), {undefined, 0}, Actors),
% ...

find_highest_actor(ActorBase) -> 
    fun(ActorID, Acc) -> 
        find_highest_actor(ActorBase, ActorID, Acc)
    end.

find_highest_actor(ActorBase, <<ActorBase:8/binary, EpochCntr:32/integer>>=A, {_HighestActor, MaxEpoch}) when
      EpochCntr > MaxEpoch ->
    {A, EpochCntr};
find_highest_actor(ActorBase, <<ActorBase:8/binary, EpochCntr:32/integer>>, {_HighestActor, MaxEpoch}=Acc) when
      EpochCntr =< MaxEpoch -> 
    %% TODO: (sdc) This clause may not even be necessary! When the
    %% EpochCntr isn't greater, we just pass along the existing
    %% highest actor.
    Acc;
%% Since an actor without an epoch is lower than an actor with one,
%% this means in the unmatched case, the ActorBase passes through as
%% the highest actor, and the epoch (of zero) passes through too.
find_highest_actor(_ActorBase, _, Acc) ->
    Acc.
```

Or even simpler:

``` erlang
find_highest_actor(ActorBase) -> 
    fun(<<MatchingBase:8/binary, EpochCntr:32/integer>>=A, {_HighestActor, MaxEpoch}) 
        when MatchingBase == ActorBase andalso 
             EpochCntr > MaxEpoch -> 
               {A, EpochCntr};
        (_, Acc) ->
               %% Since an actor without an epoch is lower than an
               %% actor with one, this means in the unmatched case,
               %% the ActorBase passes through as the highest actor,
               %% and the epoch (of zero) passes through too.
               %%
               %% NB (sdc): this also covers where the ActorBase
               %% doesn't match, i.e. it is some other actor entirely.
               Acc
    end.
```

I guess the reason I like to try to identify things that look like folds an reimplement them as such is it forces me to find ways to make things more succinct. In this instance, I found some of those clauses aren't even needed -- two are handled by the implementation of `lists:foldl/3`, and one is a subset of a catch-all case.

russelldb replied:
Ok

russelldb replied:
Just re-wrote it as you suggest, indeed it is more concise and there were redundant clauses. Thank you!

seancribbs commented on src/riak_kv_vnode_status_mgr.erl:
```diff
@@ -0,0 +1,394 @@
+          vnode_pid :: undefined | pid()
+         }).
+
+-type status() :: [proplists:property()] | [].
```
The empty list is not strictly necessary in this type. Also, is there any case where you will have bare atoms in the list? If not, `orddict:orddict()` might be better all around.

russelldb replied:
I don't know, it is inherited.

seancribbs commented on src/riak_kv_vnode_status_mgr.erl:
```diff
@@ -0,0 +1,394 @@
+         }).
+
+-type status() :: [proplists:property()] | [].
+-type init_args() :: [init_arg()].
```
Might be safer to use a tuple instead of a list, then the ordering is explicit and dialyzer can find incorrect orderings easier.

``` erlang
-type init_args() :: {pid(), non_neg_integer()}.
```

russelldb replied:
I can do that? This is cargo-culted from all other Riak gen_servers

seancribbs commented on src/riak_kv_vnode_status_mgr.erl:
```diff
@@ -0,0 +1,394 @@
+get_vnodeid_and_counter(Pid, LeaseSize) when is_integer(LeaseSize),
+                                          LeaseSize > 0,
+                                          LeaseSize =< ?MAX_CNTR ->
+    %% @TODO decide on a sane timeout for getting a vnodeid/cntr pair
```
Yes, you should probably use `infinity` timeout here, or otherwise have a safety mechanism that ensures you don't have false negatives. There could be a race with the disk and the vnode deciding to give up on the counter process.

russelldb replied:
Can you describe this race? Seems that the time out causes the vnode to crash, which causes the manager to crash. Then the vnode comes back up, and we start over.

seancribbs replied:
So, if the vnode does a strict match, that makes sense. Thanks for clarifying.

seancribbs commented on src/riak_kv_vnode_status_mgr.erl:
```diff
@@ -0,0 +1,394 @@
+%%--------------------------------------------------------------------
+-spec clear_vnodeid(pid()) -> {ok, cleared}.
+clear_vnodeid(Pid) ->
+    gen_server:call(Pid, clear).
```
In `clear_vnodeid`, `status`, and `stop`, consider whether the five-second default timeout makes sense at all. It may be better as `infinity`.

seancribbs commented on src/riak_kv_vnode_status_mgr.erl:
```diff
@@ -0,0 +1,394 @@
+            {VnodeId, Status2} = assign_vnodeid(erlang:now(),
+                                                riak_core_nodeid:get(),
+                                                Status),
+            {0, LeaseSize, VnodeId, replace(counter, LeaseSize, Status2)};
```
Like I mentioned in the type specs above, if no duplicates or bare keys are allowed or used in the status lists, it might as well be an `orddict`. Then you can just use `orddict:store/3`.

seancribbs commented on src/riak_kv_vnode_status_mgr.erl:
```diff
@@ -0,0 +1,394 @@
+read_vnode_status(File) ->
+    case file:consult(File) of
+        {ok, [Status]} when is_list(Status) ->
+            {ok, proplists:delete(version, Status)};
```
Should you be checking the version read from disk to ensure it's the same?

seancribbs replied:
Also, if you switch to `orddict` for the status, you'll want to use `from_list` and `erase` to accomplish the same thing as `proplists:delete`.

russelldb replied:
WRT to checking version, maybe should check that it is 1 or 2? Since older vnode statuses can be read by the new manager.

seancribbs commented on src/riak_kv_vnode_status_mgr.erl:
```diff
@@ -0,0 +1,394 @@
+-endif.
+
+
+replace_test() ->
```
This test will become unnecessary if you use `orddict`.

seancribbs commented on src/riak_kv_vnode_status_mgr.erl:
```diff
@@ -0,0 +1,394 @@
+-ifdef(TEST).
+
+-ifdef(EQC).
+%% prop_monotonic_counter() ->
```
Did this get moved to the separate EQC module? If so, no need to keep it here.

seancribbs commented on src/riak_object.erl:
```diff
@@ -75,7 +75,8 @@
 
 -export([new/3, new/4, ensure_robject/1, ancestors/1, reconcile/2, equal/2]).
--export([increment_vclock/2, increment_vclock/3]).
+-export([increment_vclock/2, increment_vclock/3, prune_vclock/3, vclock_descends/2, all_actors/1]).
```
I love that you increased the encapsulation here. We should look to move more things from the vnode into the riak_object.

russelldb commented:
I rebased against 2.0 before it moves too far away. I addressed your comments. Sadly I can't push to this branch now. Instead I opened a new PR and ref'd this one from riak_kv#1053
